File: java-to-go-projects/08-cli-task-runner/Main.java
package taskrunner;

import java.util.*;

/**
 * CLI entry point for the task runner.
 *
 * Usage:
 *   java Main <task-name>           — Run a specific task (+ its deps)
 *   java Main --all                 — Run all tasks in dependency order
 *   java Main --list                — List all tasks
 *   java Main --file=custom.json <task>  — Use a custom tasks file
 *
 * Go structure:
 *   main.go                   — CLI parsing, orchestration
 *   taskrunner/task.go        — Task struct
 *   taskrunner/parser.go      — ParseFile, Validate, cycle detection
 *   taskrunner/executor.go    — Run (with deps), RunAll, topological sort
 *                                Uses os/exec.Command for shell execution
 *
 * Rust structure:
 *   main.rs                   — CLI parsing
 *   task.rs                   — Task struct, Deserialize
 *   parser.rs                 — parse_file, validate, cycle detection (DFS)
 *   executor.rs               — run, run_all, topo_sort
 *                                Uses std::process::Command for shell execution
 */
public class Main {

    public static void main(String[] args) throws Exception {
        String tasksFile = "tasks.json";
        String targetTask = null;
        boolean runAll = false;
        boolean listOnly = false;

        for (String arg : args) {
            if (arg.startsWith("--file=")) {
                tasksFile = arg.split("=", 2)[1];
            } else if (arg.equals("--all")) {
                runAll = true;
            } else if (arg.equals("--list")) {
                listOnly = true;
            } else {
                targetTask = arg;
            }
        }

        Map<String, Task> tasks = Parser.parseFile(tasksFile);

        if (listOnly) {
            System.out.println("Available tasks:");
            tasks.values().forEach(t ->
                    System.out.printf("  %-15s %s  deps=%s%n",
                            t.getName(), t.getCommand(), t.getDependsOn()));
            return;
        }

        Executor executor = new Executor(tasks);

        if (runAll) {
            executor.runAll();
        } else if (targetTask != null) {
            executor.run(targetTask);
        } else {
            System.err.println("Usage: task-runner <task-name> | --all | --list");
            System.exit(1);
        }
    }
}

File: java-to-go-projects/08-cli-task-runner/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
)

type Task struct {
	Name      string   `json:"name"`
	Command   string   `json:"command"`
	DependsOn []string `json:"depends_on"`
	WorkDir   string   `json:"work_dir,omitempty"`
}

// parseFile reads a tasks.json file into a map of name -> Task.
func parseFile(filename string) (map[string]Task, error) {
	_ = json.Unmarshal
	panic("todo")
}

// validate checks that all deps exist and there are no circular dependencies (DFS).
func validate(tasks map[string]Task) error {
	panic("todo")
}

// topologicalSort returns task names in dependency order.
func topologicalSort(tasks map[string]Task) []string {
	panic("todo")
}

// runTask executes a single task using os/exec. Returns error on failure.
func runTask(task Task) error {
	_ = exec.Command("sh", "-c", task.Command)
	panic("todo")
}

// runAll runs all tasks in dependency order.
func runAll(tasks map[string]Task) error {
	panic("todo")
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <task-name|--all|--list> [--file=tasks.json]\n", os.Args[0])
		os.Exit(1)
	}

	panic("todo: parse args, load tasks file, validate, run task(s)")
}
